Make EPS installable with pip

suvayu opened the pull request:
This resolves #40.

Summary:
1. EPS is now pip installable (`pip3 install --user .` or `python3 setup.py install|test|pytest`):
    - the certificate and README is included in the package
    - platform independent scripts (`eps-daemon`, `eps-rescan`) are generated by setuptools during installation
2. Better logging, more configurable and easy to manage
    - separate streams for console and logfile
    - logfile is more verbose (only by default)
    - if it is better to decouple this entirely, I can workup a patch
3. Option parsing for scripts
    - configuration file has to be mandatorily provided on the cli
    - logging is configurable with cli options

chris-belcher commented:
Thanks for the PR. I'm still reviewing. I'll have to learn a little bit about how pip works.

suvayu commented:
@chris-belcher, sure!  I'm also not an expert.

If it helps, here are some of the docs that I refered to while working on this PR:
- [`setuptools`](https://setuptools.readthedocs.io/en/latest/setuptools.html)
  - see the section on keywords
  - including data files
  - automatic script creation
- [Integration with pytest](https://docs.pytest.org/en/latest/goodpractices.html?highlight=setup.py#integrating-with-setuptools-python-setup-py-test-pytest-runner) (minor)
- For argument parsing and logging I used the `argparse` and `logging` modules from the Python standard library.

Also an important point, I do not have access to a Windows machine; so while I tried my best not to break things on Windows, I haven't been able to test it.

TychoVrahe commented:
Tried to run this on Windows machine, few minor issues:

- Fails when "C:\tmp" folder is missing and no other path is given. Had to create the folder manually.
- os.getenv('USER') does not work on Windows, resulting in log filename being set as eps-None.log. os.getenv('USERNAME') would work, but that would not work on linux. Probably best to ditch the username from the log filename. 
- The data_files (including certs) is copied to "../../etc/eps/" relatively to eps-daemon when installing with --user (without this option it copies to "../etc/eps/"), which is not good, since by default the cert folder is expected at "." relatively to eps-daemon. Had to copy the folder manually (changing the config file would work too). If changing this in the config file is to be necessary, the documentation should mention this.
- On my Windows machine, the installation destination (when using the --user option) was C:\Users\\_USERNAME_\AppData\Roaming\Python\Python36\Scripts, should be specified in readme also. This is not by default in PATH, so trying to run eps-daemon from other location wont work.

After dealing with this, it worked fine (but the testing was not really thorough, basically just tried to run this and connect electrum).

I would not say that this makes things on Windows easier for non technical users. Also not sure about making the configuration file specification mandatory.

suvayu commented:
Hi @TychoVrahe, thank you so much for testing. 

#### Considerations before implementation
`pip install [--user] eps` should lead to,
1. executables/scripts available in the path
2. data files needed for the script to run should be discoverable by the scripts
3. configuration required to get up and running should be minimal (not much beyond adding the address public keys)
4. not add additional dependencies trying to achieve 1-3.

#### Response to your comments
- > On my Windows machine, the installation destination (when using the --user option) was C:\Users\USERNAME\AppData\Roaming\Python\Python36\Scripts

  I guess that means the `--user` option does not respect `%PATH%` (point 1 above).  Could you please confirm that the scripts are installed in the correct path without the option?  Also, if you could share the exact path, it would be very helpful.

- > The data_files (including certs) is copied to "../../etc/eps/" relatively to eps-daemon when installing with --user (without this option it copies to "../etc/eps/"), which is not good, since by default the cert folder is expected at "." relatively to eps-daemon

  Hmm, are you sure about `eps-daemon` expecting it at `.`?  Here's the relevant lines in the [source](https://github.com/suvayu/electrum-personal-server/blob/1db20e5126fa96bbbf9779fae2ee9787cd28d9df/eps/server/common.py#L461-L476), on Linux this translates to `~/.local/lib/python3.6/site-packages/eps/__init__.py` -> `~/.local/`, following which the certificate file should be at `~/.local/etc/eps/certs/cert.crt`.  Without the `--user` flag, `~/.local` becomes `/usr/local/`, but everything else remains the same.

  Could you please share the exact Windows paths with and without `--user`?  It would be really helpful to resolve this issue.

- > os.getenv('USER') does not work on Windows

  I'll remove this.

- > Fails when "C:\tmp" folder is missing

  Thank you for spotting this.  I was sloppy, should have used [`tempfile`](https://docs.python.org/3/library/tempfile.html).

- > Also not sure about making the configuration file specification mandatory.

  This is required because there is no standard configuration directory on Windows.  Since the config file needs to be modified, it is not a lot to require the file path as an argument.

#### Remarks
I would be satisfied if we can reach one of these two outcomes:
1. `pip install`
2. edit config file
3. `eps-daemon -c /path/to/config.cfg`

or,
1. `pip install`
2. edit config file
3. copy scripts to your `PATH` (this would mean the user is still using a git checkout, not pypi)
4. `eps-daemon -c /path/to/config.cfg`

TychoVrahe commented:
With or without the --user option it does not seem to be working correctly. 

Without --user:
eps-daemon is in `C:\Python36\Scripts`
certs folder is in `C:\Python36\etc\eps`
and the ` get_certs` function searches for `c:\etc/eps/certs/cert.crt` and `c:\etc/eps/certs/cert.key`

With --user
eps-daemon is in `C:\Users\\_USERNAME_\AppData\Roaming\Python\Python36\Scripts`
certs folder is in `C:\Users\\_USERNAME_\AppData\Roaming\Python\etc\eps`
and the ` get_certs` function searches for `C:\Users\\_USERNAME_\AppData\Roaming\etc/eps/certs/cert.crt` and `C:\Users\\_USERNAME_\AppData\Roaming\etc/eps/certs/cert.key`

Seems like I am missing the `lib` folder on windows for some reason, and therefore I am one folder off.

(placing the certs into "." worked for me too, because the default config file specifies certs/cert.crt path, sorry about causing confusion with this)

TychoVrahe commented:
`C:\Users\USERNAME\AppData\Roaming\Python\Python36\site-packages\eps\__init__.py`

and  

`C:\Python36\Lib\site-packages\eps\__init__.py`

suvayu commented:
The paths aren't even consistent between themselves!  I'll take the "internal data" route, thanks a lot @TychoVrahe for testing and answering my queries :).

TychoVrahe commented:
Great, its working.

Without --user option:
eps-daemon: `C:\Python36\Scripts`
top level `__init__.py`: `C:\Python36\Lib\site-packages\eps`
certs: `C:\Python36\Lib\site-packages\eps\certs`

With --user option:
eps-daemon: `C:\Users\USERNAME\AppData\Roaming\Python\Python36\Scripts`
top level `__init__.py`: `C:\Users\USERNAME\AppData\Roaming\Python\Python36\site-packages\eps`
certs: `C:\Users\USERNAME\AppData\Roaming\Python\Python36\site-packages\eps\certs`

chris-belcher commented on eps/server/common.py:
```diff
@@ -41,55 +40,52 @@
-    debug_fd.write(line + "\n")
-    debug_fd.flush()
+def logger_config(logger, fmt=None, filename=None, logfilemode='w'):
+    formatter = logging.Formatter(fmt, datefmt='%Y-%m-%d %H:%M:%S')
```
Please add ",%f" at the end of the dateformat so that the timestamps have sub-second resolution. It's useful for debugging events that happen quicker than a second.

suvayu replied:
Will do

chris-belcher commented:
I've reviewed this. It's a good PR, I'd like to merge it.

I've left inline comments on some of the code and have some general comments:

*) I'd prefer to not use the acronym EPS in the code or the package name. That acronym can be used for many things and I don't want to use up the namespace, and acronyms are not very user friendly. The `eps-daemon` isn't used very much so doesn't have to be short like `cat` or `sed`, so it could be called `electrumpersonalserver` or `electrum-personal-server`.

*) It would be good to have a entry point scripts in the root folder too, which would make it a bit easier for developers to run the server after downloading from github. Like how Electrum wallet has the `run_electrum` script but also can be installed with pip.

*) It would be good to have a way to have the debug log information be saved to a file, but have the normal logging information outputted to stdout. This is how it works in the status quo. The debug file is useful if a user reports a bug so we can look in their debug file for more information, but the debug logs won't clutter up the user's stdout during normal use.

I intend to implement code to fix #22 and do a release because Bitcoin Core 0.17 is coming out soon, after that I'd hope to merge this PR and do another release.

Also I'll ask @Stadicus if he could try this new way of installing, because he maintains a guide for installing EPS on a raspberry pi.

suvayu commented:
Hi @chris-belcher,

Thank you for reviewing.  I'll address your inline comments in the code.

- > I'd prefer to not use the acronym EPS in the code or the package name. 

  Understood, I'll make the appropriate changes.

- > It would be good to have a entry point scripts in the root folder too

  Noted. Since the scripts don't really change with the code, I could "copy paste" the scripts generated by `pip` and add them to the repo.  The only time this would need to be updated is when someone updates the `console_scripts` option in `setup.py`. Do you think that would be an acceptable solution?

- > It would be good to have a way to have the debug log information be saved to a file, but have the normal logging information outputted to stdout

  I think that is default behaviour now. Is it behaving differently for you?  That would be a bug.

- > I intend to implement code to fix #22 and do a release because Bitcoin Core 0.17 is coming out soon, after that I'd hope to merge this PR and do another release.

  I would like to submit the `tor` PR before you make the second release.  I hope you haven't forgotten that :).  The work is ready apart from finishing touches.  I would like some user testing though.  I'll try to think of a list of things to test and post in #36 with instructions.

suvayu commented:
A summary of the fixes:

#### Naming
- All names in the source are: `electrumpersonalserver`, package name, logger name, etc
- All names that might be typed on the terminal are: `electrum-personal-server`, script names (`electrum-personal-server{,-rescan}`), package name as known to `pip` (the name we would use when uploading to pypi), etc.

#### Logging
- timestamp has millisecond resolution
- the log format is cleaner now: `level:timestamp: message`
- new toggle option to append to log file instead of overwriting
- *not a change, just clarifying*: when debug logging is enabled with `-d` or `--loglevel`, the debug messages go solely to the log file stream, the terminal output is limited to info messages.  If you reduce verbosity by setting `--loglevel` to, say, `ERROR`, then both the terminal and log file will be limited to `ERROR`.

#### For developers
Regarding how to run the scripts when editing the source, all you need to do is install **once** with `pip3 install -e /path/to/repo` (with or without `--user`), this puts the console scripts in the usual places like a regular install, but the modules are imported from the repo.  So the developer can happily update the source, and the changes will be directly visible to console scripts.  No need for any hacks or workarounds.  I have documented this in the README.

Cheers

chris-belcher commented:
Thanks for the edits.

> *    It would be good to have a way to have the debug log information be saved to a file, but have the normal logging information outputted to stdout

> I think that is default behaviour now. Is it behaving differently for you? That would be a bug.

Yes. For me when running under default conditions (`electrum-personal-server -c config.cfg`) both the terminal and log file show only INFO logging information. Me adding the command line flag `-d` or `--loglevel DEBUG` makes the terminal print out INFO and the log file have DEBUG. I'm running on XFCE Ubuntu 16.04.1 LTS.

Nit: Also print out the location of the debug log with log level INFO. So that someone knows easily where to find the log file.

suvayu commented:
Hi @chris-belcher,

Sorry for the late response, I'm travelling.

On Sat 1 Sep, 2018, 22:14 chris-belcher, <notifications@github.com> wrote:

> Yes. For me when running under default conditions (electrum-personal-server
> -c config.cfg) both the terminal and log file show only INFO logging
> information. Me adding the command line flag -d or --loglevel DEBUG makes
> the terminal print out INFO and the log file have DEBUG. I'm running on
> XFCE Ubuntu 16.04.1 LTS.
>
Sorry, since it works for you as I intended, I'm not sure what I should do.
Should I change the default logging mode to debug? I guess that makes sense
since Electrum Personal Server is still in beta.

Nit: Also print out the location of the debug log with log level INFO. So
> that someone knows easily where to find the log file.
>
Very good point! I'll add this.

Since I don't have my laptop with me, give me a day or two to implement
this. I hope that's fine.

Cheers,


-- 

-- 
Suvayu

Open source is the future. It sets us free.

chris-belcher commented:
> Sorry, since it works for you as I intended, I'm not sure what I should do. Should I change the default logging mode to debug? I guess that makes sense since Electrum Personal Server is still in beta.

I'd say yes. The terminal should only show INFO and the log file should show DEBUG. Bitcoin Core does it in a similar way (except it doesnt really have INFO)

> Since I don't have my laptop with me, give me a day or two to implement this. I hope that's fine.

No worries. In your own time.

suvayu commented:
Hi @chris-belcher,

Sorry for the immense delay.  I returned from my trip this weekend.

I addressed your comments (DEBUG by default, and log file name in the logging stream), rebased on top of the latest tip, and rewrote/squashed the commits as much as possible (trying to squash a few more leads to a lot of manual conflict resolutions).  Please let me know if this is acceptable.

Cheers
